src/gen9_avc_encoder.c

```diff
@@ -3926,55 +4442,378 @@ gen9_avc_send_surface_mbenc(VADriverContextP ctx,
+        gen9_add_buffer_gpe_surface(ctx,
+                                    gpe_context,
+                                    gpe_resource,
+                                    0,
+                                    size / 4,
+                                    0,
+                                    GEN9_AVC_MBENC_MV_PREDICTOR_INDEX);
+    }
+
+    /* mb qp */
+    if (fei_param->mb_qp && (fei_param->qp != VA_INVALID_ID)) {
+        size = frame_mb_nums  + 3;
+        gpe_resource = &avc_priv_surface->res_fei_mb_qp_surface,
+        gen9_add_buffer_gpe_surface(ctx,
+                                    gpe_context,
+                                    gpe_resource,
+                                    0,
+                                    size / 4,
+                                    0,
+                                    GEN9_AVC_MBENC_MBQP_INDEX);
+    }
+
+
+    /*=== FEI distortion surface ====*/
+    size = frame_mb_nums * 48; //sizeof (VAEncFEIDistortionBufferH264Intel) == 48
+    gpe_resource = &avc_priv_surface->res_fei_vme_distortion_surface;
+    gen9_add_buffer_gpe_surface(ctx,
+                                gpe_context,
+                                gpe_resource,
+                                0,
+                                size / 4,
+                                0,
+                                GEN9_AVC_MBENC_AUX_VME_OUT_INDEX);
+
+    return;
+}
+
+static VAStatus
+gen9_avc_kernel_mbenc(VADriverContextP ctx,
+                      struct encode_state *encode_state,
+                      struct intel_encoder_context *encoder_context,
+                      bool i_frame_dist_in_use)
+{
+    struct i965_driver_data *i965 = i965_driver_data(ctx);
+    struct i965_gpe_table *gpe = &i965->gpe_table;
+    struct encoder_vme_mfc_context * vme_context = (struct encoder_vme_mfc_context *)encoder_context->vme_context;
+    struct generic_encoder_context * generic_ctx = (struct generic_encoder_context *)vme_context->generic_enc_ctx;
+    struct i965_avc_encoder_context * avc_ctx = (struct i965_avc_encoder_context *)vme_context->private_enc_ctx;
+    struct generic_enc_codec_state * generic_state = (struct generic_enc_codec_state *)vme_context->generic_enc_state;
+    struct avc_enc_state * avc_state = (struct avc_enc_state *)vme_context->private_enc_state;
+
+    struct i965_gpe_context *gpe_context;
+    struct gpe_media_object_walker_parameter media_object_walker_param;
+    struct gpe_encoder_kernel_walker_parameter kernel_walker_param;
+    unsigned int downscaled_width_in_mb, downscaled_height_in_mb;
+    int media_function = 0;
+    int kernel_idx = 0;
+    unsigned int mb_const_data_buffer_in_use = 0;
+    unsigned int mb_qp_buffer_in_use = 0;
+    unsigned int brc_enabled = 0;
+    unsigned int roi_enable = (generic_state->num_roi > 0) ? 1 : 0;
+    unsigned int dirty_roi_enable = ((generic_state->dirty_num_roi > 0) && (generic_state->frame_type == SLICE_TYPE_P) && (0));
```

Comment:
Not something I included in this patch series. I am trying to avoid any non-fei changes in this patch set to make it clean. If you need, please file a new bug on this.